### ISSUE TITLE
Is it possible to read more than one file, using some pattern?

### ISSUE THREAD

**Issue by aborruso** [duckdb/duckdb_spatial#191]
Hi,
I have a lot of file pbf in a folder. I am able to read a single file using `SELECT * FROM ST_READ('11_1076_766.pbf');`.

It seems not possibile to run something like `SELECT * FROM ST_READ('/folder/*.pbf');`, I have this error:

```
Error: IO Error: GDAL Error (4): /folder/*.pbf: No such file or directory
```

Is there another way to do it? Is there another way to merge all?

Thank you

**Comment by Maxxen** [comment]
I don't think this is possible right now, but it's actually a pretty intriguing idea. DuckDB has some reusable code used by the json/csv/parquet reader functions to handle multiple file inputs and maybe we would make it work for ST_Read too with some work.

However you can get the available files matching a pattern from within duckdb using `SELECT * FROM glob('...')` function.

**Comment by aborruso** [comment]
> I don't think this is possible right now, but it's actually a pretty intriguing idea. DuckDB has some reusable code used by the json/csv/parquet reader functions to handle multiple file inputs and maybe we would make it work for ST_Read too with some work.

Do I turn this into a feature request?

Thank you

**Comment by marklit** [comment]
It would be great to see glob support added to ``ST_READ``.

```bash
$ ls -lht ne_10m_bathymetry_*.shx
```

```
-rwxrwxrwx  1 mark  staff   124B Aug 13  2017 ne_10m_bathymetry_A_10000.shx
-rwxrwxrwx  1 mark  staff   276B Aug 13  2017 ne_10m_bathymetry_B_9000.shx
-rwxrwxrwx  1 mark  staff   220B Aug 13  2017 ne_10m_bathymetry_C_8000.shx
-rwxrwxrwx  1 mark  staff   572B Aug 13  2017 ne_10m_bathymetry_D_7000.shx
-rwxrwxrwx  1 mark  staff   4.2K Aug 13  2017 ne_10m_bathymetry_E_6000.shx
-rwxrwxrwx  1 mark  staff    15K Aug 13  2017 ne_10m_bathymetry_F_5000.shx
-rwxrwxrwx  1 mark  staff    14K Aug 13  2017 ne_10m_bathymetry_G_4000.shx
-rwxrwxrwx  1 mark  staff   5.3K Aug 13  2017 ne_10m_bathymetry_H_3000.shx
-rwxrwxrwx  1 mark  staff   1.7K Aug 13  2017 ne_10m_bathymetry_I_2000.shx
-rwxrwxrwx  1 mark  staff   1.3K Aug 13  2017 ne_10m_bathymetry_J_1000.shx
-rwxrwxrwx  1 mark  staff   4.2K Aug 13  2017 ne_10m_bathymetry_K_200.shx
-rwxrwxrwx  1 mark  staff   276B Aug 13  2017 ne_10m_bathymetry_L_0.shx
```

```bash
$ duckdb
```

```sql
SELECT COUNT(*) FROM ST_READ('ne_10m_bathymetry_*.shx');
```

```
Error: IO Error: GDAL Error (4): ne_10m_bathymetry_*.shx: No such file or directory
```

**Comment by jaanli** [comment]
+1, need this as well -- trying to combine shapefiles in this directory: https://www2.census.gov/geo/tiger/TIGER2020/PUMA20/ (to make this plot spatial: https://jaanli.github.io/american-community-survey/income)

SQL that does not work with the spatial extension's lack of support for glob syntax:

```
COPY(
SELECT * FROM '~/data/american_community_survey/TIGER2020/PUMA20/**/*.shp'
) TO '~/data/american_community_survey/shapefiles/TIGER2020/PUMA20/combined/TIGER2020_PUMA20.shp' WITH (FORMAT GDAL, DRIVER 'ESRI Shapefile');
```

Otherwise, this extension is amazing and lets me do things I could not before -- thank you!!

**Comment by yutannihilation** [comment]
A naive idea: while it sounds nice to handle multiple files using DuckDB's mechanism, probably we can just merge multiple layers on GDAL's side? More specifically, create a new layer on memory ([MEM driver](https://gdal.org/en/stable/drivers/vector/mem.html)) and copy the content of the layers one by one.

It might not be very performant, but I think it's not very frequent when the performance matters. For example,

a. the files are not very large
b. the purpose is to merge all the data so all bytes need to be parsed and copied anyway

**Comment by Maxxen** [comment]
So I've sketched an implementation of this using DuckDBs new MultiFileReader abstraction, but I don't actually know how it's supposed to work in the context of GDAL. What are the semantics of multi-file-reading multiple files that themselves contain multiple layers? How does the user say e.g, 
```
select from all my gpkg files and 
   [union all tables in all files]
   [union only the table named 'layer' in all files] 
   [union only the table named 'layer' or 'layer_2' if present] 
   [union table 'layer` in files starting with `foo_*` and table 'layer_2' in files ending with '*_bar']
```

**Comment by Maxxen** [comment]
Yeah that makes sense, I guess im overthinking it... If we can emit multiple "virtual" files (one for each layer) in the multi-file-reader, DuckDB might be able to handle the rest.

**Comment by onnimonni** [comment]
I was also hit by this today and I assumed that `ST_Read()` would accept arrays similiarly as `read_parquet()` accepts.

Anything I could help to make this possible?

**Comment by ericemc3** [comment]
Hi ! Is there a roadmap for that very useful functionality ? ;)

**Comment by aborruso** [comment]
@yutannihilation created an extension to do this https://github.com/yutannihilation/duckdb-ext-st-read-multi

**Comment by yutannihilation** [comment]
Yes! I already adversed it in the above comment :)
